### PR TITLE
Optimize performance (~3.5X faster build when images are up-to-date)

### PR DESCRIPTION
**Pull request by skovhus**
This is a second iteration of #44, where we optimize the performance of Brick when images are already built. 

In #44 we used the modified timestamps of inputs and the BUILD.yaml file to speed up `brick prepare` and `brick build`. But as git does not preserve the timestamps, we **now rely on the hash of the content of all dependency files** (inputs and BUILD.yaml file). 

Note that hashing is more expensive that comparing modified timestamps (which we also didn't do correctly in #44, but only on folders). But for all the targets at tmrow the execution time of `sha1sum` for all dependency files is roughly 10-200 ms. At least locally this is considerably faster than the overhead of buildkit (the bottleneck here is "resolve image config" and "load metadata"), which is around 1-1.5 seconds per image when everything is cached.

## Design consideration

First I did a ~/.brick-cache.json file containing all hashes for tags. But I realized that this would become out of date with the docker images once images are pruned, etc. So instead we now simply add a `LABEL` (key/value part of the Dockerfile) to the image. So we have a single source of truth. And the lookup is simple: "do we have any images built for tag X and Y with dependency hash A?".

## Timings

I got these timings locally when nothing changed and everything was already prepared and built:

    time brick -r prepare    : 48 seconds 
    time brick -r build      : 175 seconds

On this branch:

    time brick -r prepare    : 5 seconds
    time brick -r build      : 52 seconds

So when everything is up-to-date: brick build time is ~3.5X faster and prepare time is 10X faster. 🚀

## TODO

- [x] Depends on merging https://github.com/tmrowco/brick/pull/45
- [x] Test on CI and verify performance increase

### CODE REVIEW

**Comment by madsnedergaard** [comment]
# MAGIC! 🧙 

Screenshot 2020-09-01 at 10 10 08

**Comment by skovhus** [comment]
> MAGIC! 🧙

Oh, I'm happy to hear this. :D

File: brick/dockerlib.py
Comment (madsnedergaard) [inline brick/dockerlib.py:32]: Should this still be here? Seems like a left-over from testing :)

Comment (madsnedergaard) [inline brick/dockerlib.py]: Should this still be here? Seems like a left-over from testing :)

File: brick/dockerlib.py
Comment (madsnedergaard) [inline brick/dockerlib.py:34]: Just a FYI (on how great this is): I actually think the overhead is potentially way bigger - currently it takes ~10 seconds per package in a no-op run where everything should be cached!

File: brick/dockerlib.py
Comment (madsnedergaard) [inline brick/dockerlib.py:38]: Clever solution with using `labels`, seems like it avoids a lot of issues - and also makes it easier to do rollback, etc. :)

File: brick/__main__.py
Comment (madsnedergaard) [inline brick/__main__.py:206]: Good catch, didn't think about changes to the `BUILD` file :)

**Review by madsnedergaard — COMMENTED** [review]
One minor thing to look at, otherwise it looks great!

**Comment by corradio** [comment]
Let me know if you need an explicit review from me @skovhus . Else I'll let @madsnedergaard review.

File: brick/dockerlib.py
Comment (skovhus) [reply brick/dockerlib.py]: My bad. Yes, Pylint found it (I prepared to set that up in [another PR](https://github.com/tmrowco/brick/pull/48) together with other development tooling).

File: brick/dockerlib.py
Comment (skovhus) [reply brick/dockerlib.py:38]: Yeah, it turned out to be much simpler than maintaining a caching file. :)

File: brick/dockerlib.py
Comment (skovhus) [reply brick/dockerlib.py:34]: Wow, that is a lot! 

Locally an overhead of 1.5 second explains the 175-53 second speedup, as we skip building 80 images * 1.5 second. But I guess it can be much slower on CI.

**Review by madsnedergaard — APPROVED** [review]
LGTM!

**Pull request merged**
